How do you write a resilient HTTP Hyper server that does not crash with "too many open files"?

The Hello world example at https://hyper.rs/ is vulnerable to denial of service attacks if the max number of allowed open file descriptors is not high enough. I was in contact with @seanmonstar already about this and he does not think this is a security issue, so I'm posting this publicly.

Steps to reproduce:

1. Implement the Hello world example from https://hyper.rs/
2. Set a very low file descriptor limit to 50 to provoke the crash early: `ulimit -n 50`
3. Start the Hello world server: `cargo run`
4. Open another shell and attack the server with Apache bench (100 concurrent requests): `ab -c 100 -n 10000 http://localhost:3000/`

That will crash the server with an IO Error `Io(Error { repr: Os { code: 24, message: "Too many open files" } })`.

A naive solution is to just restart the server all the time with a loop:

```rust
fn main() {
    loop {
        let addr = "127.0.0.1:3000".parse().unwrap();
        let server = Http::new().bind(&addr, || Ok(Proxy)).unwrap();
        match server.run() {
            Err(e) => println!("Error: {:?}", e),
            Ok(_) =>{},
        };
    }
}
```

Which is not at all ideal because there is a downtime for a short period of time and all connections from clients are reset.

I checked the behavior of other server software, Varnish in this case. With a low file descriptor limit it just waits until it has descriptors available before accepting connections.

Can Hyper do the same? How do you run your Hyper servers in production to prevent a server crash when file descriptors run out?

I'm trying the `or_else()` future as @carllerche recommended.

Starting from:

```rust
let server = listener.incoming().for_each(move |(sock, addr)| {
    http.bind_connection(
        &handle,
        sock,
        addr,
        Proxy {
            port: port,
            upstream_port: upstream_port,
            client: client.clone(),
        },
    );
    Ok(())
});
```

Attempt 1: just insert the or_else() and see what the compiler tells us:

```rust
let server = listener.incoming().or_else(|e| {
                println!("{:?}", e);
            }).for_each(move |(sock, addr)| { ... });
```

```
error[E0277]: the trait bound `(): futures::Future` is not satisfied
   --> src/lib.rs:169:46
    |
169 |             let server = listener.incoming().or_else(|e| {
    |                                              ^^^^^^^ the trait `futures::Future` is not implemented for `()`
    |
    = note: required because of the requirements on the impl of `futures::IntoFuture` for `()`
```

I was hoping the compiler would give me a hint about the return type I have to produce in my closure, but that is not helpful. I'm not using `()` anywhere, so what is she talking about? Looking at the docs at https://docs.rs/futures/0.1.16/futures/stream/trait.Stream.html#method.or_else there is no example and the type only says I need to return an `U` which is `IntoFuture<Item = Self::Item>`.

Attempt 2: Return an empty Ok tuple as the for_each() does

```rust
let server = listener.incoming().or_else(|e| {
                println!("{:?}", e);
                Ok(())
            }).for_each(move |(sock, addr)| { ... });
```

```
error[E0271]: type mismatch resolving `<std::result::Result<(), _> as futures::IntoFuture>::Item == (tokio_core::net::TcpStream, std::net::SocketAddr)`
   --> src/lib.rs:169:46
    |
169 |             let server = listener.incoming().or_else(|e| {
    |                                              ^^^^^^^ expected (), found tuple
    |
    = note: expected type `()`
               found type `(tokio_core::net::TcpStream, std::net::SocketAddr)`
```

Attempt 3: Return an Err:

```rust
let server = listener.incoming().or_else(|e| {
                println!("{:?}", e);
                Err(e)
            }).for_each(move |(sock, addr)| { ... });
```

At least it compiles!!!

But it does not solve the problem: returning an error here bubbles up and crashes my server as before. With the only difference of the additional print statement.

Attempt 4: Return an empty future, assuming it does nothing and `Incoming` continues with the next connection attempt:

```rust
let server = listener.incoming().or_else(|e| -> futures::future::Empty<_, std::io::Error> {
    println!("{:?}", e);
    futures::future::empty()
}).for_each(move |(sock, addr)| { ... });
```

That compiles, but as soon as the first IO error happens the server does not respond anymore. Looking at the docs: https://docs.rs/futures/0.1.16/futures/future/struct.Empty.html it says "A future which is never resolved.". Aha, so that is probably blocking my server. So this is not really an Empty future and should be renamed to "AlwaysBlockingDoingNothing".

Attempt 5: Let's try the `or_else()` after the `for_each()`:

```rust
let server = listener.incoming().for_each(...).or_else(|e| -> Result<_> {
    println!("{:?}", e);
    Ok(())
});
```

This compiles, but does not swallow the error. The server still crashes except for the additional print statement.

At this point I'm running out of ideas. How can I swallow the IO error and make the `incoming` future continue?

--- a/examples/hello.rs
+++ b/examples/hello.rs
@@ -19,7 +19,10 @@ fn main() {
             .with_body(PHRASE))
     }));
 
-    let server = Http::new().bind(&addr, new_service).unwrap();
+    let server = Http::new()
+        .sleep_on_errors(true)
+        .bind(&addr, new_service)
+        .unwrap();
     println!("Listening on http://{} with 1 thread.", server.local_addr().unwrap());
     server.run().unwrap();
 }

--- a/src/server/mod.rs
+++ b/src/server/mod.rs
@@ -664,6 +664,7 @@ impl Stream for AddrIncoming {
     type Error = ::std::io::Error;
 
     fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        // Check if a previous timeout is active that was set by IO errors.
         if let Some(ref mut to) = self.timeout {
             match to.poll().expect("timeout never fails") {
                 Async::Ready(_) => {}
@@ -682,8 +683,13 @@ impl Stream for AddrIncoming {
                     return Ok(Async::Ready(Some(AddrStream::new(socket, addr))));
                 },
                 Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => return Ok(Async::NotReady),
-                Err(ref e) if connection_error(e) => continue,
-                Err(e) => {
+                Err(ref e) if self.sleep_on_errors => {
+                    // Connection errors can be ignored directly, continue by
+                    // accepting the next request.
+                    if connection_error(e) {
+                        continue;
+                    }
+                    // Sleep 10ms.
                     let delay = ::std::time::Duration::from_millis(10);
                     debug!("Accept error: {}. Sleeping {:?}...",
                         e, delay);
@@ -698,7 +704,8 @@ impl Stream for AddrIncoming {
                             return Ok(Async::NotReady);
                         }
                     }
-                }
+                },
+                Err(e) => return Err(e),
             }
         }
     }

--- a/src/server/mod.rs
+++ b/src/server/mod.rs
@@ -57,6 +57,7 @@ pub struct Http<B = ::Chunk> {
     max_buf_size: Option<usize>,
     keep_alive: bool,
     pipeline: bool,
+    sleep_on_errors: bool,
     _marker: PhantomData<B>,
 }
 
@@ -102,6 +103,9 @@ pub struct AddrIncoming {
     addr: SocketAddr,
     keep_alive_timeout: Option<Duration>,
     listener: TcpListener,
+    handle: Handle,
+    sleep_on_errors: bool,
+    timeout: Option<Timeout>,
 }
 
 /// A future binding a connection with a Service.
@@ -144,6 +148,7 @@ impl<B: AsRef<[u8]> + 'static> Http<B> {
             keep_alive: true,
             max_buf_size: None,
             pipeline: false,
+            sleep_on_errors: false,
             _marker: PhantomData,
         }
     }
@@ -169,6 +174,18 @@ impl<B: AsRef<[u8]> + 'static> Http<B> {
     /// Default is false.
     pub fn pipeline(&mut self, enabled: bool) -> &mut Self {
         self.pipeline = enabled;
+        self
+    }
+
+    /// Swallow connection accept errors. Instead of passing up IO errors when
+    /// the server is under heavy load the errors will be ignored. Some
+    /// connection accept errors (like "connection reset") can be ignored, some
+    /// (like "too many files open") may consume 100% CPU and a timout of 10ms
+    /// is used in that case.
+    ///
+    /// Default is false.
+    pub fn sleep_on_errors(&mut self, enabled: bool) -> &mut Self {
+        self.sleep_on_errors = enabled;
         self
     }
 
@@ -225,7 +242,7 @@ impl<B: AsRef<[u8]> + 'static> Http<B> {
               Bd: Stream<Item=B, Error=::Error>,
     {
         let listener = TcpListener::bind(addr, &handle)?;
-        let mut incoming = AddrIncoming::new(listener)?;
+        let mut incoming = AddrIncoming::new(listener, handle.clone(), self.sleep_on_errors)?;
         if self.keep_alive {
             incoming.set_keepalive(Some(Duration::from_secs(90)));
         }
@@ -248,6 +265,7 @@ impl<B: AsRef<[u8]> + 'static> Http<B> {
                 keep_alive: self.keep_alive,
                 max_buf_size: self.max_buf_size,
                 pipeline: self.pipeline,
+                sleep_on_errors: self.sleep_on_errors,
                 _marker: PhantomData,
             },
         }
@@ -394,7 +412,7 @@ impl<S, B> Server<S, B>
 
         let handle = reactor.handle();
 
-        let mut incoming = AddrIncoming::new(listener)?;
+        let mut incoming = AddrIncoming::new(listener, handle.clone(), protocol.sleep_on_errors)?;
 
         if protocol.keep_alive {
             incoming.set_keepalive(Some(Duration::from_secs(90)));
@@ -619,11 +637,14 @@ mod unnameable {
 // ===== impl AddrIncoming =====
 
 impl AddrIncoming {
-    fn new(listener: TcpListener) -> io::Result<AddrIncoming> {
+    fn new(listener: TcpListener, handle: Handle, sleep_on_errors: bool) -> io::Result<AddrIncoming> {
          Ok(AddrIncoming {
             addr: listener.local_addr()?,
             keep_alive_timeout: None,
             listener: listener,
+            handle: handle,
+            sleep_on_errors: sleep_on_errors,
+            timeout: None,
         })
     }
 
@@ -643,6 +664,13 @@ impl Stream for AddrIncoming {
     type Error = ::std::io::Error;
 
     fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        if let Some(ref mut to) = self.timeout {
+            match to.poll().expect("timeout never fails") {
+                Async::Ready(_) => {}
+                Async::NotReady => return Ok(Async::NotReady),
+            }
+        }
+        self.timeout = None;
         loop {
             match self.listener.accept() {
                 Ok((socket, addr)) => {
@@ -654,10 +682,39 @@ impl Stream for AddrIncoming {
                     return Ok(Async::Ready(Some(AddrStream::new(socket, addr))));
                 },
                 Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => return Ok(Async::NotReady),
-                Err(e) => return Err(e),
+                Err(ref e) if connection_error(e) => continue,
+                Err(e) => {
+                    let delay = ::std::time::Duration::from_millis(10);
+                    debug!("Accept error: {}. Sleeping {:?}...",
+                        e, delay);
+                    let mut timeout = Timeout::new(delay, &self.handle)
+                        .expect("can always set a timeout");
+                    let result = timeout.poll()
+                        .expect("timeout never fails");
+                    match result {
+                        Async::Ready(()) => continue,
+                        Async::NotReady => {
+                            self.timeout = Some(timeout);
+                            return Ok(Async::NotReady);
+                        }
+                    }
+                }
             }
         }
     }
+}
+
+/// This function defines errors that are per-connection. Which basically
+/// means that if we get this error from `accept()` system call it means
+/// next connection might be ready to be accepted.
+///
+/// All other errors will incur a timeout before next `accept()` is performed.
+/// The timeout is useful to handle resource exhaustion errors like ENFILE
+/// and EMFILE. Otherwise, could enter into tight loop.
+fn connection_error(e: &io::Error) -> bool {
+    e.kind() == io::ErrorKind::ConnectionRefused ||
+    e.kind() == io::ErrorKind::ConnectionAborted ||
+    e.kind() == io::ErrorKind::ConnectionReset
 }
 
 mod addr_stream {

